### PR TITLE
feat: Calling new API version when creating clusters

### PR DESCRIPTION
**Pull request by AgustinBettati**
## Description

Link to any related issue(s): CLOUDP-257162

This PR adjusts the create operation to start using the new API exclusively, but still preserving the possibility of creating clusters defined in the old schema structure.
- If numShards is > 1 additional replication spec object are added in the request.
- If disk size gb is defined at the root level, it is propagated to the hardware spec level.
- New expand functions supporting new API model. Existing expand functions are preserved as they are currently being used for the update operation which was not modified.

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [x] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR. A migration guide must be created or updated if the new feature will go in a major version.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR. A migration guide must be created or updated.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have read the [contributing guides](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/contributing/README.md)
- [x] I have checked that this change does not generate any credentials and that **they are NOT accidentally logged anywhere**.
- [ ] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [ ] I have added any necessary documentation (if appropriate)
- [x] I have run make fmt and formatted my code
- [x] If changes include deprecations or removals, I defined an isolated PR with a relevant title as it will be used in the auto-generated changelog.
- [x] If changes include removal or addition of 3rd party GitHub actions, I updated our internal document. Reach out to the APIx Integration slack channel to get access to the internal document.

### CODE REVIEW

File: internal/service/advancedcluster/model_advanced_cluster.go [hunk 1]
```diff
@@ -15,7 +15,8 @@ import (
 	"github.com/mongodb/terraform-provider-mongodbatlas/internal/common/constant"
 	"github.com/mongodb/terraform-provider-mongodbatlas/internal/common/conversion"
 	"github.com/spf13/cast"
-	"go.mongodb.org/atlas-sdk/v20231115014/admin"
+	admin20231115 "go.mongodb.org/atlas-sdk/v20231115014/admin"
+	"go.mongodb.org/atlas-sdk/v20240530001/admin"
```
Comment (AgustinBettati) [inline internal/service/advancedcluster/model_advanced_cluster.go:19]: preview version will be called with `admin`, while currrent stable version will use alias `admin20231115` as this will become the fixed old sdk version.

**Comment by AgustinBettati** [comment]
Keeping an eye on `TestAccClusterAdvancedCluster_multicloudSharded` as creation seems to be flaky when using new API, sometimes fails with `DISK_SIZE_INVALID_FOR_AZURE` while currently in nightly runs (using old API) this issue does not appear.

File: internal/service/advancedcluster/model_advanced_cluster.go [hunk 2]
```diff
@@ -773,35 +775,66 @@ func expandLabelSliceFromSetSchema(d *schema.ResourceData) ([]admin.ComponentLab
...
 		if !ok || tfMap == nil {
 			continue
 		}
-		apiObject := expandAdvancedReplicationSpec(tfMap)
+		apiObject := expandAdvancedReplicationSpec(tfMap, rootDiskSizeGB)
+		apiObjects = append(apiObjects, *apiObject)
+
+		// handles adding additional replication spec objects if legacy num_shards attribute is being used and greater than 1
+		numShards := tfMap["num_shards"].(int)
+		for range numShards - 1 {
```
Comment (lantoli) [inline internal/service/advancedcluster/model_advanced_cluster.go:790]: nice!

File: internal/service/advancedcluster/model_advanced_cluster.go [hunk 3]
```diff
@@ -864,6 +897,10 @@ func expandRegionConfigSpec(tfList []any, providerName string) *admin.DedicatedH
 	if v, ok := tfMap["node_count"]; ok {
 		apiObject.NodeCount = conversion.Pointer(v.(int))
 	}
+
+	// Once disk_size_gb schema attribute is added at this level, we will check if defined in config and prioritize over value defined at root level (deprecated and to be reomved)
```
Comment (EspenAlbert) [inline internal/service/advancedcluster/model_advanced_cluster.go]: last word typo: `reomved` -> `removed`

File: internal/service/advancedcluster/model_advanced_cluster.go [hunk 4]
```diff
@@ -773,35 +775,66 @@ func expandLabelSliceFromSetSchema(d *schema.ResourceData) ([]admin.ComponentLab
...
 }
 
-func expandAdvancedReplicationSpec(tfMap map[string]any) *admin.ReplicationSpec {
-	apiObject := &admin.ReplicationSpec{
+func expandAdvancedReplicationSpec(tfMap map[string]any, rootDiskSizeGB *float64) *admin.ReplicationSpec20240710 {
+	apiObject := &admin.ReplicationSpec20240710{
+		ZoneName:      conversion.StringPtr(tfMap["zone_name"].(string)),
+		RegionConfigs: expandRegionConfigs(tfMap["region_configs"].([]any), rootDiskSizeGB),
+	}
+	// here we will populate id value using external_id value (relevant for update)
```
Comment (maastha) [inline internal/service/advancedcluster/model_advanced_cluster.go]: nit: the comments are a bit confusing, can we mark relevant ones as `TODO`s for now so it's easy to keep track

**Review by maastha — APPROVED** [review]
minor comment, LGTM

File: internal/service/advancedcluster/model_advanced_cluster.go [hunk 5]
```diff
@@ -773,35 +775,66 @@ func expandLabelSliceFromSetSchema(d *schema.ResourceData) ([]admin.ComponentLab
...
+	if len(tfList) == 0 {
+		return nil
+	}
+	var apiObjects []admin20231115.ReplicationSpec
+	for _, tfMapRaw := range tfList {
+		tfMap, ok := tfMapRaw.(map[string]any)
+		if !ok || tfMap == nil {
+			continue
+		}
+		apiObject := expandAdvancedReplicationSpecOldSDK(tfMap)
```
Comment (lantoli) [inline internal/service/advancedcluster/model_advanced_cluster.go:807]: knit: it can happen that tfList contains elements but they're not ok so we'll end up returning a non-nil zero-size list.

if we want always to return nil for empty slice, we might delete the if condition from the beginning and having something like this at the end:

if apiObjects == nil {
return nil
}
return &apiObjects

File: internal/service/advancedcluster/model_advanced_cluster.go [hunk 6]
```diff
@@ -773,35 +775,66 @@ func expandLabelSliceFromSetSchema(d *schema.ResourceData) ([]admin.ComponentLab
...
 	}
 	if tfMap["id"].(string) != "" {
 		apiObject.Id = conversion.StringPtr(tfMap["id"].(string))
 	}
 	return apiObject
 }
 
-func expandRegionConfigs(tfList []any) *[]admin.CloudRegionConfig {
+func expandRegionConfigs(tfList []any, rootDiskSizeGB *float64) *[]admin.CloudRegionConfig {
 	if len(tfList) == 0 {
```
Comment (lantoli) [inline internal/service/advancedcluster/model_advanced_cluster.go]: same comment as above

File: internal/service/advancedcluster/model_sdk_version_conversion.go [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	admin20231115 "go.mongodb.org/atlas-sdk/v20231115014/admin"
+	"go.mongodb.org/atlas-sdk/v20240530001/admin"
+)
+
+// Conversions from one SDK model version to another are used to avoid duplicating our flatten/expand conversion functions.
+// - These functions must not contain any business logic.
+// - All will be removed once we rely on a single API version.
+
+func convertTagsToLatest(tags *[]admin20231115.ResourceTag) *[]admin.ResourceTag {
```
Comment (lantoli) [inline internal/service/advancedcluster/model_sdk_version_conversion.go:13]: we might want a companion test file with unit tests

File: internal/service/advancedcluster/model_sdk_version_conversion.go [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+		return nil
+	}
+	return &admin20231115.BiConnector{
+		Enabled:        biconnector.Enabled,
+		ReadPreference: biconnector.ReadPreference,
+	}
+}
+
+func convertLabelSliceToOldSDK(slice []admin.ComponentLabel, err diag.Diagnostics) ([]admin20231115.ComponentLabel, diag.Diagnostics) {
+	if err != nil {
```
Comment (lantoli) [inline internal/service/advancedcluster/model_sdk_version_conversion.go:40]: seems strange to have err param just to check if it's set. i suppose it's a convenience for the caller

File: internal/service/advancedcluster/resource_advanced_cluster.go [hunk 9]
```diff
@@ -940,11 +945,11 @@ func getUpgradeRequest(d *schema.ResourceData) *admin.LegacyAtlasTenantClusterUp
...
 func updateAdvancedCluster(
 	ctx context.Context,
-	connV2 *admin.APIClient,
-	request *admin.AdvancedClusterDescription,
+	connV2 *admin20231115.APIClient,
+	request *admin20231115.AdvancedClusterDescription,
 	projectID, name string,
 	timeout time.Duration,
-) (*admin.AdvancedClusterDescription, *http.Response, error) {
+) (*admin20231115.AdvancedClusterDescription, *http.Response, error) {
```
Comment (lantoli) [inline internal/service/advancedcluster/resource_advanced_cluster.go:952]: will tests go in a different PR?

**Review by lantoli — APPROVED** [review]

File: internal/service/advancedcluster/model_advanced_cluster.go [hunk 10]
```diff
@@ -773,35 +775,66 @@ func expandLabelSliceFromSetSchema(d *schema.ResourceData) ([]admin.ComponentLab
...
 		tfMap, ok := tfMapRaw.(map[string]any)
 		if !ok || tfMap == nil {
 			continue
 		}
-		apiObject := expandAdvancedReplicationSpec(tfMap)
+		apiObject := expandAdvancedReplicationSpec(tfMap, rootDiskSizeGB)
+		apiObjects = append(apiObjects, *apiObject)
+
+		// handles adding additional replication spec objects if legacy num_shards attribute is being used and greater than 1
+		numShards := tfMap["num_shards"].(int)
```
Comment (EspenAlbert) [inline internal/service/advancedcluster/model_advanced_cluster.go:789]: Would be nice to see some test examples of how the added replication_specs looks when `num_shards > 1` ,I think the replication IDs might need to be added for the update calls, but that probably goes in the next "update" PR?

**Review by EspenAlbert — APPROVED** [review]
LGTM. Struggled a bit to follow the logic of the different SDK models. Maybe a test would help clarify 👍

File: internal/service/advancedcluster/model_advanced_cluster.go (hunk 4 above)
Comment (AgustinBettati) [reply internal/service/advancedcluster/model_advanced_cluster.go]: makes sense, adjusted

File: internal/service/advancedcluster/model_advanced_cluster.go (hunk 5 above)
Comment (AgustinBettati) [reply internal/service/advancedcluster/model_advanced_cluster.go:807]: nice catch, adjusted this logic

File: internal/service/advancedcluster/model_advanced_cluster.go (hunk 6 above)
Comment (AgustinBettati) [reply internal/service/advancedcluster/model_advanced_cluster.go]: adjusted

File: internal/service/advancedcluster/model_sdk_version_conversion.go (hunk 7 above)
Comment (AgustinBettati) [reply internal/service/advancedcluster/model_sdk_version_conversion.go:13]: had my doubts on value of tests in this case as there is no business logic other than passing values from one model to another

File: internal/service/advancedcluster/model_sdk_version_conversion.go (hunk 8 above)
Comment (AgustinBettati) [reply internal/service/advancedcluster/model_sdk_version_conversion.go:40]: correct, essentially enables this code from call perspective:
```
labels, err := convertLabelSliceToOldSDK(expandLabelSliceFromSetSchema(d))
if err != nil {
	return err
}
cluster.Labels = &labels

File: internal/service/advancedcluster/resource_advanced_cluster.go (hunk 9 above)
Comment (AgustinBettati) [reply internal/service/advancedcluster/resource_advanced_cluster.go:952]: yes, to define a working acceptance test with symmetric sharded cluster and asymmetric sharded cluster using the new schema first the read operation needs to be adjusted to call the new API (coming in separate PR).

File: internal/service/advancedcluster/model_advanced_cluster.go (hunk 10 above)
Comment (AgustinBettati) [reply internal/service/advancedcluster/model_advanced_cluster.go:789]: as mentioned [here](https://github.com/mongodb/terraform-provider-mongodbatlas/pull/2363#discussion_r1654889746), we will need to adjust the read operation to consume the new API, and then support the update as well. As these operations are implemented we can start including tests capturing the new scenarios.
